File: pset6/vigenere/vigenere.py
from cs50 import get_string

import sys


def main() -> int:
    if len(sys.argv) == 2:
        key = Key(sys.argv[1])
        phrase = get_string('plaintext:  ')

        ciphertext = encode(phrase, key)

        print(f'ciphertext: {ciphertext}')
        sys.exit(0)
    else:
        print("Usage: vigenere.py key")
        sys.exit(1)


def encode(phrase, key) -> str:
    result = ''
    for c in phrase:
        offset = key.offset()

        if c.isupper():
            letter = shift(c, offset)
            key.increment()
        elif c.islower():
            letter = shift(c.upper(), offset).lower()
            key.increment()
        else:
            letter = c

        result += letter
    return result


def shift(char, offset) -> chr:
    shifted = chr(ord(char) + offset)

    if ord(shifted) > ord('Z'):
        shifted = chr(ord(shifted) - 26)

    return shifted


class Key:
    def __init__(self, value):
        if not value.isalpha():
            print('Key must contain only letters')
            sys.exit(1)
        self.value = value.upper()
        self.index = 0

    def increment(self) -> None:
        self.index = (self.index + 1) % len(self.value)

    def offset(self) -> int:
        return ord(self.value[self.index]) - ord('A')


if __name__ == "__main__":
    main()

File: pset2/vigenere/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int increment_key(int index, int length)
{
    index += 1;
    return index % length;
}

char shift(char letter, int offset)
{
    letter += offset;
    if (letter > 'Z')
    {
        letter = ('A' - 1) + (letter % 'Z');

    }
    return letter;
}

char *encode(string phrase, string key)
{
    int keyLength = strlen(key);
    int keyIndex = 0;

    for (int i = 0; i < strlen(phrase); i++)
    {
        int offset = toupper(key[keyIndex]) - 'A';

        if (islower(phrase[i]))
        {
            phrase[i] = toupper(phrase[i]);
            phrase[i] = shift(phrase[i], offset);
            phrase[i] = tolower(phrase[i]);
            keyIndex = increment_key(keyIndex, keyLength);
        }
        else if (isupper(phrase[i]))
        {
            phrase[i] = shift(phrase[i], offset);
            keyIndex = increment_key(keyIndex, keyLength);
        }
    }
    return phrase;
}

bool contains_only_letters(string phrase)
{
    for (int i; i < strlen(phrase); i++)
    {
        if (!isalpha(phrase[i]))
        {
            return false;
        }
    }
    return true;
}

bool valid_args(int argc, string argv[])
{
    if (argc == 2 && contains_only_letters(argv[1]))
    {
        return true;
    }
    return false;
}

int main(int argc, string argv[])
{
    if (valid_args(argc, argv))
    {
        string key = argv[1];
        string phrase = get_string("plaintext:");
        phrase = encode(phrase, key);
        printf("ciphertext: %s\n", phrase);
    }
    else
    {
        printf("invalid arguments\n");
        return 1;
    }
    return 0;
}